### PR TITLE
feat: support png, gif, webp

### PR DESCRIPTION
**Pull request by ZuzooVn**
# Checklist:

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

# Description

### PNg
- Before
Screenshot 2024-09-04 at 6 39 46 AM
- After
Screenshot 2024-09-04 at 6 40 07 AM

### GIF, WEBP
- Before
Screenshot 2024-09-04 at 6 40 37 AM
- After

https://github.com/user-attachments/assets/88b9eda8-4305-41da-bda3-24d7c26c5f12

## Type of Change

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Comment by xuzuodong** [comment]
Since animated images do not require cropping, we need to ensure that even if the user uploads a non-square image, it will be displayed without stretching. 

image

File: web/app/components/base/app-icon-picker/Uploader.tsx [hunk 1]
```diff
@@ -8,12 +8,12 @@ import classNames from 'classnames'
 
 import { ImagePlus } from '../icons/src/vender/line/images'
 import { useDraggableUploader } from './hooks'
+import { isAnimatedImage } from './utils'
 import { ALLOW_FILE_EXTENSIONS } from '@/types/app'
 
 type UploaderProps = {
   className?: string
-  onImageCropped?: (tempUrl: string, croppedAreaPixels: Area, fileName: string) => void
```
Comment (xuzuodong) [inline web/app/components/base/app-icon-picker/Uploader.tsx:15]: This callback function now would be better named as `onUpload` or something IMO.

File: web/app/components/base/app-icon-picker/utils.ts [hunk 2]
```diff
@@ -93,6 +93,10 @@ export default async function getCroppedImg(
         resolve(file)
       else
         reject(new Error('Could not create a blob'))
-    }, 'image/jpeg')
+    }, 'image/png')
   })
 }
+
+export function isAnimatedImage(file: { name: string }) {
```
Comment (xuzuodong) [inline web/app/components/base/app-icon-picker/utils.ts]: I think in most cases `webp` could be a static image, maybe we can use a better way to determain if an uploaded image is animated? 

I asked ChatGPT, and here's the answer maybe we can refer to:

```javascript
function isAnimatedWebP(file) {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.onload = function(event) {
      const data = event.target.result;
      const view = new DataView(data);
      
      // Checking for the WebP RIFF header.
      if (view.getUint32(0, false) !== 0x52494646 /* 'RIFF' */ ||
          view.getUint32(8, false) !== 0x57454250 /* 'WEBP' */) {
        reject(new Error("Not a WebP file"));
        return;
      }
      
      // Checking for the presence of the 'ANIM' chunk.
      let offset = 12;
      while (offset < view.byteLength) {
        const chunkType = view.getUint32(offset, false);
        const chunkSize = view.getUint32(offset + 4, true);
        
        // 'ANIM'
        if (chunkType === 0x414E494D /* 'ANIM' */) {
          resolve(true);
          return;
        }
        
        // Move to the next chunk (chunkSize is padded to even number of bytes)
        offset += 8 + chunkSize + (chunkSize % 2);
      }
      resolve(false); // No animation chunk found
    };
    reader.onerror = function() {
      reject(new Error("File reading failed"));
    };
    reader.readAsArrayBuffer(file);
  });
}
```

Or maybe we can arbitrarily remove `webp` determination here :)

File: web/app/components/base/app-icon-picker/Uploader.tsx [hunk 3]
```diff
@@ -8,12 +8,12 @@ import classNames from 'classnames'
...
-}
+  onImageCropped?: (tempUrl: string, croppedAreaPixels: Area, fileName: string, file?: File) => void }
```
Comment (xuzuodong) [inline web/app/components/base/app-icon-picker/Uploader.tsx]:
```suggestion
  onImageCropped?: (tempUrl: string, croppedAreaPixels: Area, fileName: string, file?: File) => void
}
```

File: web/app/components/base/app-icon-picker/index.tsx [hunk 4]
```diff
@@ -69,9 +69,9 @@ const AppIconPicker: FC<AppIconPickerProps> = ({
     },
   })
 
-  const [imageCropInfo, setImageCropInfo] = useState<{ tempUrl: string; croppedAreaPixels: Area; fileName: string }>()
-  const handleImageCropped = async (tempUrl: string, croppedAreaPixels: Area, fileName: string) => {
-    setImageCropInfo({ tempUrl, croppedAreaPixels, fileName })
+  const [imageCropInfo, setImageCropInfo] = useState<{ tempUrl: string; croppedAreaPixels: Area; fileName: string; file?: File }>()
+  const handleImageCropped = async (tempUrl: string, croppedAreaPixels: Area, fileName: string, file?: File) => {
+    setImageCropInfo({ tempUrl, croppedAreaPixels, fileName, file })
   }
```
Comment (xuzuodong) [inline web/app/components/base/app-icon-picker/index.tsx:80]: `handleImageCropped` could therefore rename to `handleUpload`, and `imageCropInfo` to `uploadImageInfo`

**Comment by ZuzooVn** [comment]
thank @xuzuodong , I will fix it asap

File: web/app/components/base/app-icon-picker/Uploader.tsx (hunk 3 above)
Comment (xuzuodong) [reply web/app/components/base/app-icon-picker/Uploader.tsx]: @ZuzooVn Can you please revert this?

**Comment by xuzuodong** [comment]
The implementation for determining animated .webp images looks good to me.

Could you also review the other minor issues mentioned above?

**Comment by ZuzooVn** [comment]
> The implementation for determining animated .webp images looks good to me.
> 
> Could you also review the other minor issues mentioned above?

@xuzuodong I did it

**Comment by xuzuodong** [comment]
@ZuzooVn Uploading an animated image will raise an error now. 

https://github.com/user-attachments/assets/e7f2d937-bcda-4aa8-aaba-5187dfa3d4c3

**Comment by ZuzooVn** [comment]
@xuzuodong I can not reproduce this issue. Could you please send me that image?

**Comment by xuzuodong** [comment]
Here's the webp image:

[animated-webp-supported.webp.zip](https://github.com/user-attachments/files/17340459/animated-webp-supported.webp.zip)


Although I don't think there's anything special about this image; the error is clearly caused by accessing `file` property from an `undefined` value. If you check `web/app/components/base/app-icon-picker/index.tsx`, there's a lot of TypeScript error: 

https://github.com/user-attachments/assets/49eb9447-af9a-42c4-bc67-7eba02c49630

**Comment by crazywoola** [comment]
Hello can you resolve those comments.

**Review by crazywoola — APPROVED** [review]

**Pull request merged**